### PR TITLE
BRAVE-250

### PR DESCRIPTION
**Pull request by gnoremac**
-Added tests to show InheritableThreadLocals will leak across threads

These tests will currently fail. I don't know if it's possible to provide a reliable fix without getting rid of the InheritableThreadLocal. Even if the objects are cloned they will not work in Executors that reuse threads.

A better approach would be to maintain a stack something like

```
class WrappedSpanLocal {
    Span span
    WrappedSpanLocal parent;
}
```

and use ThreadLocals. This would involve updating the BraveExecutors that pass context between the threads to not only pass server spans but client and local spans as well which would break backwards compatibility.

I will continue to think of another solution.

### CODE REVIEW

**Comment by adriancole** [comment]
cc @schlosna I'm pretty busy next couple days. in case you aren't mind reviewing this?

**Comment by adriancole** [comment]
It looks like you are actually testing the behavior of inheritable thread locals, right? We'd expect inheritable thread locals to propagate to threads.

Can you rephrase your tests on the actual use case you are trying to get to work? or maybe summarize the reason why the inherited state causes a problem? (I know there are a number of reasons why many don't prefer inheritable, but we need to pin this down with a use case)

**Comment by gnoremac** [comment]
Sorry I totally forgot about this issue. Your right my test is really testing thread locals. I guess the issue I'm having is Spans are immutable, most users by default will use a executor that reuses threads which results in weird behaviour as the span moves between threads.

I'm not sure I even understand the use case for InheritableServerClientAndLocalSpanState as it seems like it's very unsafe due to the thread locals. 

Maybe this issue is more of a documentation update issue than a code fix issue as it seems like it's working as intended.

**Comment by adriancole** [comment]
latest brave has a type called `CurrentTraceContext` which can be easily customized to not use inheritable thread locals
